### PR TITLE
[connector/datadog] Use the native OTel ingest API in APM stats

### PR DESCRIPTION
**Pull request by songy23**
**Description:**
Add a feature gate `connector.datadogconnector.NativeIngest` that enables datadog connector to use the new native OTel API in APM stats computation. It is disabled by default.

Follow-up of https://github.com/DataDog/datadog-agent/pull/23503.

### CODE REVIEW

File: go.mod
Comment (songy23) [inline go.mod:272]: Using a commit version for now, probably need to wait for the official release before merging this PR.

File: connector/datadogconnector/connector_native.go
Comment (dineshg13) [inline connector/datadogconnector/connector_native.go:45]: rename to statsout ?

File: connector/datadogconnector/connector_native.go
Comment (dineshg13) [inline connector/datadogconnector/connector_native.go:60]: why are we doing it ?

**Review by dineshg13 — APPROVED** [review]
Looks good overall have few nit comments

File: connector/datadogconnector/connector_native.go
Comment (songy23) [reply connector/datadogconnector/connector_native.go:60]: This is a copy of https://github.com/open-telemetry/opentelemetry-collector-contrib/blame/main/connector/datadogconnector/connector.go#L69 and was added in https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/9a34aea049669c76d716885d347c96c5f6e3e3dc#diff-63ca951f32af68231d95ef502d34657b930abb3d1a6ffaa8db864a28f3a17d6aR51

cc @mx-psi should have more context

File: connector/datadogconnector/connector_native.go
Comment (songy23) [reply connector/datadogconnector/connector_native.go]: Sure

File: go.mod
Comment (mx-psi) [reply go.mod:272]: Blocking merge based on this comment

**Review by mx-psi — CHANGES_REQUESTED** [review]

File: connector/datadogconnector/connector_native.go
Comment (mx-psi) [reply connector/datadogconnector/connector_native.go:60]: As the code stands, if we enable this, we may have duplicate metrics from the connector and exporter with the exact same name and tags. We can add support for this, but we need to fix that first.

File: go.mod
Comment (songy23) [reply go.mod]: This is now unblocked with https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33726

File: exporter/datadogexporter/integrationtest/integration_test.go
Comment (mx-psi) [inline exporter/datadogexporter/integrationtest/integration_test.go:171]:
```suggestion
	previousVal := datadogconnector.NativeIngestFeatureGate.IsEnabled()
	err := featuregate.GlobalRegistry().Set(datadogconnector.NativeIngestFeatureGate.ID(), true)
	require.NoError(t, err)
	defer func() {
		err = featuregate.GlobalRegistry().Set(datadogconnector.NativeIngestFeatureGate.ID(), previousVal)
		require.NoError(t, err)
	}()
```

File: connector/datadogconnector/factory.go
Comment (mx-psi) [inline connector/datadogconnector/factory.go:56]: Should this log mention the feature gate to nudge people to opt in into the feature? As it stands it can be a bit confusing ("legacy" implies "not recommended" to me, but I don't have any actionable stuff to do when I see this log message)

Comment (songy23) [reply connector/datadogconnector/factory.go]: Good point, I added the feature gate in this message

**Review by mx-psi — APPROVED** [review]

**Comment by arielvalentin** [comment]
@songy23 do you all have any benchmarks or details about the potential performance benefits of using native ingest API?

I want to know if it may help improve the overall performance of our collector pipelines.

**Comment by songy23** [comment]
@arielvalentin if your traces have lots of attributes (especially when you have high cardinality of [container tags](resource_attributes_as_container_tags)), the new API has better throughput. If your traces are fairly small (with few attributes), the throughput / performance of old and new APIs are roughly on-par.

**Pull request merged**
